src/main/java/com/google/sps/servlets/CalendarServlet.java

```diff
@@ -13,28 +29,95 @@
  * Servlet for getting or creating the Calendar page for each group
  */
 @WebServlet("/calendar")
-public class CalendarServlet extends HttpServlet {
+public class CalendarServlet extends AbstractEventsServlet {
+  private final String SCOPE_TYPE = "default";
+  private final String USER_CALENDAR_PERMISSIONS = "reader";
 
   /**
    * Display the Calendar page for each group.
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    SoyFileSet sfs = SoyFileSet
-        .builder()
-        .add(new File("../../src/main/java/templates/calendar.soy"))
-        .build();
-    SoyTofu tofu = sfs.compileToTofu();
-
-    String out = tofu.newRenderer("templates.calendar.calendarPage").render();
-    response.getWriter().println(out);
+    ImmutableList<ImmutableMap<String, String>> groupsList = 
+        ServletUtils.getGroupsList(ServletUtils.getUserEntity(request.getUserPrincipal().getName()));
+
+    String groupId = getParameter(request, GROUP_ID_PROPERTY);
+
+    if (groupsList.isEmpty()) {
+      groupId = "";
+    } else if (Strings.isNullOrEmpty(groupId)) {
+      groupId = groupsList.get(0).get(GROUP_ID_PROPERTY);
+    }
+
+    String calendarId = "";
+    if (!Strings.isNullOrEmpty(groupId)) {
+      try {
+        calendarId = ServletUtils.getGroupProperty(groupId, GROUP_CALENDARID_PROPERTY);
+      } catch (Exception entityError) {
+        ServletUtils.printBadRequestError(response, ENTITY_ERROR_MESSAGE);
+        return;
+      }
+    }
+
+    if(Strings.isNullOrEmpty(calendarId)) {
+      ServletUtils.printBadRequestError(response, ENTITY_ERROR_MESSAGE);
+      return;
+    }
```

Comment:
sorry I didn't realize this until using the deployed website, but I don't think you need this if statement anymore because if the groupId is empty (if the user hasn't joined any groups) then the calendarId is also left empty, so the "no-groups-error" in calendar.soy is never displayed because this error gets displayed first